app/main.py

```diff
@@ -2,5 +2,33 @@
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    cache_data = {}
+
+    def wrapper(*args) -> Callable:
+        if args in cache_data:
+            print("Getting from cache")
+            return cache_data[args]
+        print("Calculating new result")
+        result = func(*args)
+        cache_data[args] = result
+        return result
+
+    return wrapper
+
+
+@cache
+def long_time_func(a: int, b: int, c: int) -> int:
+    return (a ** b ** c) % (a * c)
+
+
+@cache
+def long_time_func_2(numbers_tuple: tuple, power: int) -> list:
```

Comment:
The function `long_time_func_2` takes a tuple and an integer as arguments. Ensure that the tuple contains only immutable elements, as mutable elements like lists cannot be used as dictionary keys for caching.